### PR TITLE
M3-1309 attach file new ticket

### PR DESCRIPTION
**Pull request by Jskobos**
## Purpose

Add attachments during ticket creation. Note that this involves creating the ticket first, then adding an attachment as these can't currently be combined.

## Notes
* Refactor existing ticket attachment logic into components for reuse in the support ticket drawer
* Adds recompose as a dependency (technically already used as a sub-dependency for some other import)

### Add
- User can now add an attachment when creating a support ticket

### CODE REVIEW

File: src/features/Support/ticketUtils.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export const reshapeFiles = (files: FileList) => {
+  const reshapedFiles = [];
+
+  /* tslint:disable-next-line */
+  for (let i = 0; i < files.length; i++) {
```
Comment (rjstires) [inline src/features/Support/ticketUtils.ts]: Could have just been a Map right?

`return files.map((file, idx) => ({ file, name: file.name, uploading: false, uploaded: false }));`

File: src/features/Support/TicketAttachmentList.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      </Grid>
+    </Grid>
+  )
+};
+
+const styled = withStyles(styles);
+
+const enhanced = compose<CombinedProps, Props>(
+  styled,
+  withState('showMoreAttachments', 'toggle', false),
```
Comment (rjstires) [inline src/features/Support/TicketAttachmentList.tsx]: Can be with [withStateHandlers](https://github.com/acdlite/recompose/blob/master/docs/API.md#withstatehandlers).

File: src/features/Support/TicketAttachmentList.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+type CombinedProps = Props & ToggleProps & WithStyles<ClassNames>;
+
+export const addIconsToAttachments = (attachments: string[] = []) => {
+  return attachments.map((attachment, idx) => {
+    // try to find a file extension
+    const lastDotIndex = attachment.lastIndexOf('.');
+    const ext = attachment.slice(lastDotIndex + 1);
+    if (ext) {
+      if (['jpg', 'jpeg', 'png', 'bmp', 'tiff'].includes(ext.toLowerCase())) {
```
Comment (rjstires) [inline src/features/Support/TicketAttachmentList.tsx]: Minor, minor nitpick. Ever iteration of the map we recreate the static list of extensions. We can hoist that to the body of the addIconsToAttachments.

File: src/features/Support/SupportTicketDetail/AttachmentError.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+const AttachmentError: React.StatelessComponent<CombinedProps> = (props) => {
+  const { fileName, reason } = props;
+  return (
+    <Notice error
+      text={getText(fileName, reason)}
+    />
+  );
+};
+
+const styled = withStyles(styles, { withTheme: true });
```
Comment (rjstires) [inline src/features/Support/SupportTicketDetail/AttachmentError.tsx]: Update your snippet. The option withTheme, when true, will supply the entire theme object to the wrapped component.

File: src/features/Support/SupportTicketDetail/AttachmentError.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+
+import { StyleRulesCallback, Theme, withStyles, WithStyles } from 'src/components/core/styles';
+import Notice from 'src/components/Notice';
+
+type ClassNames = 'root';
+
+const styles: StyleRulesCallback<ClassNames> = (theme: Theme) => ({
```
Comment (rjstires) [inline src/features/Support/SupportTicketDetail/AttachmentError.tsx]: Update your snipper. Theme is no longer required as global typing has been resolved.

File: src/features/Support/SupportTicketDetail/AttachmentError.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+
+import { StyleRulesCallback, Theme, withStyles, WithStyles } from 'src/components/core/styles';
```
Comment (rjstires) [inline src/features/Support/SupportTicketDetail/AttachmentError.tsx]: Update your snippet. Theme is no longer required as the theme has been properly augmented.

File: src/features/Support/AttachFileForm.tsx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                </Grid>
+              </React.Fragment>
+            )
+        ))}
+      </React.Fragment>
+    );
+  }
+};
+
+const styled = withStyles(styles, { withTheme: true });
```
Comment (rjstires) [inline src/features/Support/AttachFileForm.tsx]: Get rid of withTheme.

File: src/features/Support/AttachFileForm.tsx [hunk 8]
```diff
@@ -0,0 +1,179 @@
+import * as React from 'react';
+
+import InputAdornment from 'src/components/core/InputAdornment';
+import { StyleRulesCallback, Theme, withStyles, WithStyles } from 'src/components/core/styles';
```
Comment (rjstires) [inline src/features/Support/AttachFileForm.tsx]: You dont need to import Theme (see AttachmentError for more detailed explanation).

**Comment by rjstires** [comment]
Tested happy path and sad path, works great. There's a few nit picks, but nothing significant. Well done, as always, Jared.

File: src/features/Support/AttachFileForm.tsx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    if (this.inputRef.current) {
+      this.inputRef.current.click();
+    }
+  }
+
+  removeFile = (e: React.MouseEvent<HTMLButtonElement | HTMLDivElement>) => {
+    if (!e.target) { return; }
+    const aidx = e.currentTarget.getAttribute('data-file-idx');
+    if (!aidx) { return; }
+    const idx = +aidx;
```
Comment (martinmckenna) [inline src/features/Support/AttachFileForm.tsx]: not a huge fan of these types of variable names. Makes it hard to follow along.

File: src/features/Support/AttachFileForm.tsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                        startAdornment:
+                        <InputAdornment position="end">
+                          <CloudUpload />
+                        </InputAdornment>,
+                        endAdornment:
+                          inlineDisplay &&
+                            <InputAdornment
+                              onClick={this.removeFile}
+                              position="end"
+                              data-file-idx={idx}
```
Comment (martinmckenna) [inline src/features/Support/AttachFileForm.tsx]: imo, we should move away from this pattern of relying on `idx` attributes. I think instead of creating these attributes, we should just create a new component and pass the handler to it, so that the child component can invoke `removeFile()` and pass the args it needs.

See TagsPanelItem as an example

Comment (martinmckenna) [reply src/features/Support/AttachFileForm.tsx]: Also this button prob needs a `cursor: pointer` style

File: src/features/Support/ticketUtils.ts (hunk 1 above)
Comment (rjstires) [reply src/features/Support/ticketUtils.ts]: That's silly. Carry on sir.
